src/mpid/ch4/shm/ipc/src/ipc_win.c

```diff
@@ -79,23 +78,14 @@ int MPIDI_IPC_mpi_win_create_hook(MPIR_Win * win)
     /* Determine submodule availability and IPC type based on buffer type.
      * If no submodule is available, skip initialization; if any submodule
      * is on, we need always exchange in case remote buffer can be shared by IPC. */
-    MPL_pointer_attr_t attr;
-    MPL_gpu_query_pointer_attr(win->base, &attr);
-
-#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
-    if (MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE != -1) {
-        if (attr.type != MPL_GPU_POINTER_DEV)
-            ipc_type = MPIDI_SHM_IPC_TYPE__XPMEM;
-    } else {
-        goto fn_exit;
-    }
-#else
-    goto fn_exit;
-#endif
+    MPIDI_IPC_mem_attr_t attr;
+    MPIDI_IPC_get_mem_attr(win->base, ULONG_MAX, shm_comm_ptr->rank, shm_comm_ptr, &attr);
 
```

Comment:
Why is it `ULONG_MAX`? Should it be `win->size`? Besides, we can safely exit if no submodule is enabled, without exchanging data.  Here is a summary of the expected behaviors:
1. If all submodules are disabled at configure time or runtime, the process can exit locally without exchanging data.
2. If any submodule is enabled, all process should join the collective exchange with its local memory info (handle, size, type)
3. After exchanging, if any remote process has a different IPC type comparing with the local process, skip shm optimization; but we should continue if the remote process is a zero buffer.

We need to update the comment too. The code might not be very straightforward. I'd prefer to allow mixed-type case after this PR. Steps 1 and 2 should be unchanged, just need update step 3.

Reply:
ULONG_MAX is to force xpmem and gpu also for small buffers so that the IPC selection can still happen in the get_mem_attr function.

Reply:
Probably that function should be renames `MPIDI_IPC_get_ipc_attr` instead of `get_mem_attr` as it also takes the rank and comm to decide whether IPC is needed or not.

Reply:
I still cannot see the reason why `ULONG_MAX` is right. Can you explain more?

Reply:
Actually we should not check the rank condition in `get_mem_attr`.  It is incorrect to skip XPMEM get mem_handle when rank == my_rank in window creation. We always get the mem_handle. In P2P, we skip it when the receiver rank == my_rank. This logic should be handled in win creation and P2P, separately, not inside `get_mem_attr`

Reply:
We don't skip the get_mem_attr in window creation because in that case we are passing rank from the comm itself so the two will match. For point to point we need to obey to the XPMEM threshold to enable/disable xpmem transport. For windows we always use xpmem so I need to set the size high to void query function to return IPC_TYPE__NONE

Reply:
what about we check rank != comm->rank outside get_mem_attr and pass a threshold to get_mem_attr as parameter? This way we can change the threshold for pt2pt and rma.